src/codegen/riscv64/macro-assembler-riscv64.cc

```diff
@@ -2103,102 +2143,67 @@ void TurboAssembler::LoadFPRImmediate(FPURegister dst, uint64_t src) {
-  BlockTrampolinePoolScope block_trampoline_pool(this);
+void TurboAssembler::CompareI(Register rd, Register rs, const Operand& rt,
+                              Condition cond) {
   switch (cond) {
-    case cc_always:
-      mv(rd, zero_reg);
-      break;
     case eq:
-      if (rs == zero_reg) {
-        if (rt.is_reg()) {
-          LoadZeroIfConditionZero(rd, rt.rm());
-        } else if (rt.immediate() == 0) {
-          mv(rd, zero_reg);
-        }
-      } else if (IsZero(rt)) {
-        LoadZeroIfConditionZero(rd, rs);
-      } else {
-        Sub64(t6, rs, rt);
-        LoadZeroIfConditionZero(rd, t6);
-      }
+      Seq(rd, rs, rt);
       break;
     case ne:
-      if (rs == zero_reg) {
-        if (rt.is_reg()) {
-          LoadZeroIfConditionNotZero(rd, rt.rm());
-        } else if (rt.immediate() != 0) {
-          mv(rd, zero_reg);
-        }
-      } else if (IsZero(rt)) {
-        LoadZeroIfConditionNotZero(rd, rs);
-      } else {
-        Sub64(t6, rs, rt);
-        LoadZeroIfConditionNotZero(rd, t6);
-      }
+      Sne(rd, rs, rt);
       break;
 
     // Signed comparison.
     case greater:
-      Sgt(t6, rs, rt);
-      LoadZeroIfConditionNotZero(rd, t6);
+      Sgt(rd, rs, rt);
       break;
     case greater_equal:
-      Sge(t6, rs, rt);
-      LoadZeroIfConditionNotZero(rd, t6);
-      // rs >= rt
+      Sge(rd, rs, rt);  // rs >= rt
       break;
     case less:
-      Slt(t6, rs, rt);
-      LoadZeroIfConditionNotZero(rd, t6);
-      // rs < rt
+      Slt(rd, rs, rt);  // rs < rt
       break;
     case less_equal:
-      Sle(t6, rs, rt);
-      LoadZeroIfConditionNotZero(rd, t6);
-      // rs <= rt
+      Sle(rd, rs, rt);  // rs <= rt
       break;
 
     // Unsigned comparison.
     case Ugreater:
-      Sgtu(t6, rs, rt);
-      LoadZeroIfConditionNotZero(rd, t6);
-      // rs > rt
+      Sgtu(rd, rs, rt);  // rs > rt
       break;
-
     case Ugreater_equal:
-      Sgeu(t6, rs, rt);
-      LoadZeroIfConditionNotZero(rd, t6);
-      // rs >= rt
+      Sgeu(rd, rs, rt);  // rs >= rt
       break;
     case Uless:
-      Sltu(t6, rs, rt);
-      LoadZeroIfConditionNotZero(rd, t6);
-      // rs < rt
+      Sltu(rd, rs, rt);  // rs < rt
       break;
     case Uless_equal:
-      Sleu(t6, rs, rt);
-      LoadZeroIfConditionNotZero(rd, t6);
-      // rs <= rt
+      Sleu(rd, rs, rt);  // rs <= rt
+      break;
+    case cc_always:
+      UNREACHABLE();
       break;
     default:
       UNREACHABLE();
   }
 }
 
-// dest <- (condition != 0 ? zero : dest), which is eqvuivalent to dest <-
-// condition == 0 ? dest : zero
+// dest <- (condition != 0 ? zero : dest)
 void TurboAssembler::LoadZeroIfConditionNotZero(Register dest,
                                                 Register condition) {
-  Seleqz(dest, dest, condition);
+  UseScratchRegisterScope temps(this);
+  Register scratch = temps.Acquire();
+  seqz(scratch, condition);
+  mul(dest, dest, scratch);
```

Comment:
Depending on the latency of the multiply instruction, we may want to consider an alternative here. You could instead do:
```
  seqz(scratch, condition);
  neg(scratch, scratch); // 0 is still 0, 1 becomes all 1s 
  and(dest, dest, scratch);
```

Reply:
Indeed, for HiFive U540, the above sequence (2 cycles for 2 instructions) is more efficient than using `mul` (` instruction but 5 cycles). Changed to the new sequence.

> The pipeline has a peak execution rate of one instruction per clock cycle, and is fully bypassed so that most instructions have a one-cycle result latency. There are several exceptions:
> • LW has a two-cycle result latency, assuming a cache hit.
> • LH, LHU, LB, and LBU have a three-cycle result latency, assuming a cache hit.
> • CSR reads have a three-cycle result latency.
> • MUL, MULH, MULHU, and MULHSU have a 5-cycle result latency.
> • DIV, DIVU, REM, and REMU have between a 2-cycle and 65-cycle result latency, depending on the operand values